### PR TITLE
6639 easting northing

### PR DESCRIPTION
**Pull request by serge-gaia**
* Created Easting/Northing biosys specific schema tag instead of using latitude/longitude.
* Refactored the way the `ObservationSchema` was extracting geometry by creating a new class `GeometryParser` just for that.
* The new `GeometryParser` has a reverse method than can convert geometry to data.
* Created two API POST endpoint for the conversion of data <-> geometry.
POST utils/geometry-to-data/record/{record-pk}  
and 
POST utils/data-to-geometry/record/{record-pk}
* Tests for all that.
* Fixed a bug where the project datum wasn't used when creating record without datum.
* Removed a legacy admin view for uploading record from spreadsheet.

### CODE REVIEW

File: biosys/apps/main/migrations/0006_auto_20170619_1503.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
```
Comment (serge-gaia) [inline biosys/apps/main/migrations/0006_auto_20170619_1503.py:1]: That was a missing migrations from last sprint (added more choices)

File: biosys/apps/main/utils_data_package.py [hunk 2]
```diff
@@ -596,145 +592,33 @@ def find_observation_date_field_or_throw(schema):
...
-            else:
-                return field
-        msg = "The schema doesn't include a required longitude field. " \
-              "It must have a field named {} or tagged with biosys type {}" \
-              "Alternatively you can add a foreign key the site code to extract the geometry from it. " \
-              "To create site code foreign key, add the following to you schema:{}". \
-            format(ObservationSchema.LONGITUDE_FIELD_NAME,
-                   BiosysSchema.LONGITUDE_TYPE_NAME,
-                   ObservationSchema.SITE_CODE_FOREIGN_KEY_EXAMPLE)
-        raise ObservationSchemaError(msg)
```
Comment (serge-gaia) [inline biosys/apps/main/utils_data_package.py:685]: all the logic to extract geometry has been put in the new `GeometryParser` class below

File: biosys/apps/main/utils_data_package.py [hunk 3]
```diff
@@ -965,18 +1078,3 @@ def to_workbook(self):
...
-            headers = stream.readline().rstrip('\n').split(',')
-            values = jsontableschema.compat.csv_reader(stream)
-            schema = jsontableschema.infer(headers, values, row_limit=row_limit)
-            fp.write(six.u(json.dumps(schema, indent=2, ensure_ascii=False)))
-
-
-def infer_csvs(path, row_limit=0):
-    for filename in listdir(path):
-        if filename.endswith('.csv'):
-            infer_csv(join(path, filename), join(path, filename) + '.json', row_limit)
```
Comment (serge-gaia) [inline biosys/apps/main/utils_data_package.py:987]: removed unused and probably buggy code.

File: biosys/apps/main/views.py [hunk 4]
```diff
@@ -1,153 +1,8 @@
...
-                        len(records), dataset
-                    )
-                    messages.success(self.request, msg)
-            else:
-                src_file.delete()
-                msg = '<br>'.join(errors)
-                messages.error(self.request, msg)
-                return HttpResponseRedirect(error_url)
-
-        return super(UploadDataSetView, self).form_valid(form)
```
Comment (serge-gaia) [inline biosys/apps/main/views.py:153]: legacy admin View

File: requirements.txt [hunk 5]
```diff
@@ -5,15 +5,19 @@ django-bootstrap3==6.2.2
...
 Unipath==1.1
 six==1.10
+python-dateutil==2.6.0
+
+# frictionless tableschema and datapackage
+# TODO: update to 1.0.0 when released. Problem: it breaks a lot of things.
+jsontableschema==0.6.5
+# force a version of tabulator (dependency of jsontableschema) to avoid error when upgrading
+tabulator==0.5.0
+datapackage==0.8.1
```
Comment (serge-gaia) [inline requirements.txt:21]: Nothing changed here, just added comments.

File: biosys/apps/main/api/uploaders.py [hunk 6]
```diff
@@ -194,7 +194,8 @@ def _create_record(self, row, counter):
                 tz = self.dataset.project.timezone or timezone.get_current_timezone()
                 record.datetime = timezone.make_aware(observation_date, tz)
                 # geometry
-                geometry = self.schema.cast_geometry(row, default_srid=MODEL_SRID)
+                default_srid = self.dataset.project.datum or MODEL_SRID
+                geometry = self.schema.cast_geometry(row, default_srid=default_srid)
```
Comment (tony-gaia) [inline biosys/apps/main/api/uploaders.py]: Maybe here do like line 141 of validators.py

i.e.: `self.schema.cast_geometry(data, default_srid=self.default_srid or MODEL_SRID)`

File: biosys/apps/main/api/views.py [hunk 7]
```diff
@@ -513,3 +513,66 @@ def get(self, request, *args, **kwargs):
...
+    output = OUTPUT_GEOMETRY  # set it the url
+
+    def to_geometry(self, dataset, record_data):
+        try:
+            schema = dataset.schema
+            default_srid = dataset.project.datum or constants.MODEL_SRID
+            geom_parser = schema.geometry_parser
+            geometry = geom_parser.from_record_to_geometry(record_data, default_srid=default_srid)
+            # we output in WGS84
+            geometry.transform(4326)
```
Comment (tony-gaia) [inline biosys/apps/main/api/views.py]: use a const

File: biosys/apps/main/utils_data_package.py [hunk 8]
```diff
@@ -921,6 +754,286 @@ def cast_species_name_id(self, record):
...
+
+        # some post validations.
+        # we need at least one method to get the geometry.
+        if all([
+            not self.site_code_field,
+            not self.latitude_field,
+            not self.longitude_field,
+            not self.easting_field,
+            not self.northing_field
+        ]):
```
Comment (tony-gaia) [inline biosys/apps/main/utils_data_package.py]: Woudln't:
```
 if not any([
          self.site_code_field,
          self.latitude_field,
          self.longitude_field,
          self.easting_field,
          self.northing_field
     ]):
```

Make more sense?

File: biosys/apps/main/utils_data_package.py [hunk 9]
```diff
@@ -921,6 +754,286 @@ def cast_species_name_id(self, record):
...
+            if self.easting_field:
+                record[self.easting_field.name] = point.x
+            if self.northing_field:
+                record[self.northing_field.name] = point.y
+        elif self.is_lat_long and not is_projected_srid(srid):
+            if self.longitude_field:
+                record[self.longitude_field.name] = point.x
+            if self.latitude_field:
+                record[self.latitude_field.name] = point.y
+        else:
```
Comment (tony-gaia) [inline biosys/apps/main/utils_data_package.py:996]: else pass? Why not just remove block, or is this a TODO?

File: biosys/apps/main/api/uploaders.py (hunk 6 above)
Comment (serge-gaia) [reply biosys/apps/main/api/uploaders.py]: I'm not sure which one is more readable. You think number 2?

File: biosys/apps/main/api/views.py (hunk 7 above)
Comment (serge-gaia) [reply biosys/apps/main/api/views.py]: Will do

File: biosys/apps/main/utils_data_package.py (hunk 8 above)
Comment (serge-gaia) [reply biosys/apps/main/utils_data_package.py]: Indeed. You're the smart one :-)

File: biosys/apps/main/utils_data_package.py (hunk 9 above)
Comment (serge-gaia) [reply biosys/apps/main/utils_data_package.py:996]: It was just to put a comment as a warning but I'll put a logging message.

**Comment by serge-gaia** [comment]
Made changes from CR

**Pull request merged**